File: 1008-construct-binary-search-tree-from-preorder-traversal/1008-construct-binary-search-tree-from-preorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode bstFromPreorder(int[] preorder) {
        if (preorder.length == 0){
            return null;
        }
        TreeNode root = new TreeNode(preorder[0]);
        for(int i=1; i < preorder.length; i++){
            helper(root,preorder[i]);
        }
        return root;
    }
    
    public TreeNode helper(TreeNode root, int val){
        
        if(root == null){
            return new TreeNode(val);
        }
        else if(val < root.val){
            root.left = helper(root.left,val); 
        }
        else{
            root.right = helper(root.right,val); 
        }
        
        return root;
    }
}

File: 1008-construct-binary-search-tree-from-preorder-traversal/1008-construct-binary-search-tree-from-preorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def bstFromPreorder(self, preorder: List[int]) -> Optional[TreeNode]:
        
        if len(preorder) == 0:
            return None
        
        root = TreeNode(preorder[0])

        
        def helper(root,val):
            
            if root is None:
                return TreeNode(val)
            elif root.val > val:
                root.left = helper(root.left, val)
            else:
                root.right = helper(root.right, val)
                
            return root
        
        
        for n in preorder[1:]:
            print(n)
            helper(root, n)
        
        return root
        
        
        